### PR TITLE
Auto-register C++ ttnn operations in python

### PR DESCRIPTION
**Pull request by arakhmati**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/8835

### Problem description
Currently, registered C++ operations need to be also registered in python. Which is time-consuming and error-prone. 

### What's changed
Auto-register C++ ttnn operations in python. 
- The user still has an option t register som manually
- Golden functions can be attached manually in a very convenient way
- Deleted golden functions for backward ops from ttnn as they are not tested

### Checklist
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by johanna-rock-tt — APPROVED** [review]
falcon40b changes look good to me

**Review by cglagovichTT — APPROVED** [review]
Llama changes look good

**Review by boris-drazic — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

File: ttnn/ttnn/operations/ccl.py [hunk 1]
```diff
@@ -16,6 +16,6 @@ def _validate_input_tensors(operation_name, tensor, *args, **kwargs):
         raise RuntimeError(f"{operation_name} requires input tensors to be multi-device tensor")
 
 
-all_gather = ttnn.register_operation(validate_input_tensors=_validate_input_tensors)(
```
Comment (ayerofieiev-tt) [inline ttnn/ttnn/operations/ccl.py:19]: We need to override registration like this to add validation? Lets just move it to C++?

File: ttnn/ttnn/operations/binary.py
Comment (ayerofieiev-tt) [inline ttnn/ttnn/operations/binary.py:1]: Should rename this file to binary_golden?

**Review by ayerofieiev-tt — APPROVED** [review]
Wow, thats a pretty decent change!

**Review by eyonland — CHANGES_REQUESTED** [review]
By deleting these files we prevent the possibility of linking directly to the op don't we?

**Review by uaydonat — APPROVED** [review]

**Review by nsmithtt — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Review by AleksKnezevic — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]

**Pull request merged**